File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int waterTrapped =0;
        int left_max =0, right_max=0, i=0, j= height.length-1;
        while(i < j){
            if(height[i] < height[j]){
                left_max = Math.max(left_max, height[i]);
                waterTrapped += left_max-height[i];
                i++;
            }else{
                right_max = Math.max(right_max, height[j]);
                waterTrapped += right_max-height[j];
                j--;
            }
        }
        return waterTrapped;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left_max=INT_MIN;
        int right_max =INT_MIN;
        int water=0;
        int i=0, j=height.size()-1;
        while(i<j){
            if(height[i]<=height[j]){
                left_max=max(left_max, height[i]);
                water += left_max-height[i];
                i++;
            }else{
                right_max=max(right_max, height[j]);
                water+=right_max-height[j];
                j--;
            }
        }
        return water;
    }
};